### PR TITLE
fix: ensure role is idempotent and supports check mode

### PR DESCRIPTION
**Pull request by richm**
Cause: Role was not idempotent, and did not support check mode.

Consequence: Users could not use role to ensure idempotence or
check state.

Fix: The role has been fixed to return `"changed": true` if the
role changed the state of the system, and `false` otherwise.  This
means that some operations that used to return an error, such as
trying to snapshot a volume that already had a snapshot, or trying
to remove a snapshot that does not exist, will no longer return
an error, but will instead just report `"changed": false`.  Users
should use the `list` operation to examine state of the system.

Result: The role is idempotent and supports check mode.

Signed-off-by: Rich Megginson <rmeggins@redhat.com>

### CODE REVIEW

**Comment by richm** [comment]
[citest]

File: tasks/main.yml [hunk 1]
```diff
@@ -19,22 +19,29 @@
...
-  register: snapshot_cmd
+  register: snapshot_cmd_raw
   no_log: true
   ignore_errors: true  # noqa: ignore-errors - handled below
+  changed_when: false  # change handled below too
 
 - name: Print out response
   debug:
-    var: snapshot_cmd.stdout
+    var: snapshot_cmd_raw
```
Comment (spetrosi) [inline tasks/main.yml:29]:
```suggestion
    var: snapshot_cmd_raw.stdout
```
Should it print only stdout as it was before?

File: tasks/main.yml [hunk 2]
```diff
@@ -19,22 +19,29 @@
...
-    snapshot_facts: "{{ snapshot_cmd.stdout | from_json }}"
+    snapshot_facts: "{{ snapshot_cmd['data'] }}"
   when: snapshot_lvm_action == "list"
 
 - name: Show errors
   debug:
-    msg: "{{ snapshot_cmd.stderr }}"
-  when: snapshot_cmd.rc != 0
-  failed_when: snapshot_cmd.rc != 0
+    msg: "{{ snapshot_cmd['errors'] }}"
```
Comment (spetrosi) [inline tasks/main.yml:44]:
```suggestion
    var: snapshot_cmd['errors']
```

File: tasks/main.yml [hunk 3]
```diff
@@ -19,22 +19,29 @@
...
+  when: snapshot_cmd["return_code"] != 0
+  failed_when:
+    - snapshot_cmd_raw is failed or snapshot_cmd["return_code"] != 0
```
Comment (spetrosi) [inline tasks/main.yml:47]: Can it be on the same line with `failed_when:`?

File: tests/tests_mount.yml [hunk 4]
```diff
@@ -119,6 +138,24 @@
...
+          vars:
+            snapshot_lvm_snapset_name: snapset1
+            snapshot_lvm_action: umount
+            snapshot_lvm_vg: test_vg1
+            snapshot_lvm_lv: lv1
+            snapshot_lvm_mountpoint: "{{ test_mnt_parent ~ '/lv1_mp' }}"
+
+        - name: Assert no changes for umount
+          assert:
+            that: not snapshot_cmd["changed"]
```
Comment (spetrosi) [inline tests/tests_mount.yml:157]: Cool way to test idempotency!

File: tests/verify-role-failed.yml [hunk 5]
```diff
@@ -82,11 +82,14 @@
...
       assert:
         that:
-          - snapshot_cmd.failed
-          - snapshot_cmd.rc != 0
-          - snapshot_cmd.stdout_lines is search(__snapshot_failed_regex) or
-            snapshot_cmd.stderr_lines is search(__snapshot_failed_regex)
+          - snapshot_cmd_raw is failed
+          - snapshot_cmd["return_code"] != 0
+          - snapshot_cmd["errors"] is search(__snapshot_failed_regex)
         msg: "{{ __snapshot_failed_msg }}"
```
Comment (spetrosi) [inline tests/verify-role-failed.yml:88]:
```suggestion
        var: __snapshot_failed_msg
```

**Review by spetrosi — COMMENTED** [review]
Massive work, Rich!

File: tasks/main.yml (hunk 1 above)
Comment (richm) [reply tasks/main.yml:29]: No - would prefer to see all of the fields rather than just `stdout`

File: tests/verify-role-failed.yml (hunk 5 above)
Comment (richm) [reply tests/verify-role-failed.yml:88]: Looks like `var` is not a valid parameter for `assert`
```
    _VALID_ARGS = frozenset(('fail_msg', 'msg', 'quiet', 'success_msg', 'that'))
```
`msg` is an alias for `fail_msg`

**Pull request merged**